macrosynergy/panel/adjust_weights.py

```diff
@@ -0,0 +1,248 @@
+"""
+Implementation of adjust_weights.
+"""
+
+import numpy as np
+import pandas as pd
+from typing import List, Dict, Union, Optional, Tuple, Type, Set, Callable
+import warnings
+from numbers import Number
+from packaging import version
+from macrosynergy.management.utils import (
+    reduce_df,
+    is_valid_iso_date,
+    get_cid,
+    get_xcat,
+)
+from macrosynergy.management.simulate import make_test_df
+from macrosynergy.management.types import QuantamentalDataFrame
+
+
+PD_FUTURE_STACK = (
+    dict(future_stack=True)
+    if version.parse(pd.__version__) > version.parse("2.1.0")
+    else dict(dropna=False)
+)
+
+
+def check_missing_cids_xcats(weights, adj_zns, cids, r_xcats, r_cids):
+    """
+    Checks if there are missing cids or xcats in the input DataFrame.
+    """
+    missing_xcats = list(set([weights, adj_zns]) - set(r_xcats))
+    if missing_xcats:
+        raise ValueError(f"Missing xcats: {missing_xcats}")
+
+    missing_cids = list(set(cids) - set(r_cids))
+    if missing_cids:
+        raise ValueError(f"Missing cids: {missing_cids}")
+
+
+def check_types(
+    weights: str, adj_zns: str, method: Callable, param: Number, cids: List[str]
+):
+    """
+    Type checking for the input variables of adjust_weights.
+    """
+    for _var, _name, _type in [
+        (weights, "weights", str),
+        (adj_zns, "adj_zns", str),
+        (method, "method", Callable),
+        (param, "param", Number),
+        (cids, "cids", list),
+    ]:
+        if not isinstance(_var, _type):
+            raise TypeError(f"{_name} must be a {_type}, not {type(_var)}")
+
+    if not all(isinstance(cid, str) for cid in cids):
+        raise TypeError("`cids` must be a list of strings")
+
+
+def adjust_weights_backend(
+    df_weights_wide: pd.DataFrame,
+    df_adj_zns_wide: pd.DataFrame,
+    method: Callable,
+    param: Number,
+) -> pd.DataFrame:
+    """
+    Backend function for adjust_weights. Applies the `method` function to the weights and
+    multiplies the result by the adjustment factors, and by the parameter `param`.
+
+    Parameters
+    ----------
+    df_weights_wide : pd.DataFrame
+        DataFrame with weights in wide format.
+
+    df_adj_zns_wide : pd.DataFrame
+        DataFrame with adjustment factors in wide format.
+
+    method : Callable
+        Function that will be applied to the weights to adjust them. This function must
+
+    param : Number
+        Parameter that will be passed to the method function.
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame with the adjusted weights.
+    """
+
+    assert set(df_weights_wide.columns) == set(df_adj_zns_wide.columns)
+    cids = sorted(df_weights_wide.columns)
+    dfw_result = pd.DataFrame(index=df_weights_wide.index)
+
+    for cid in cids:
+        cid_weight = df_weights_wide[cid]
+        cid_adj_zns = df_adj_zns_wide[cid]
+
+        result = cid_weight * cid_adj_zns.apply(method) * param
+        dfw_result[cid] = result
```

Comment:
Would it not be more efficient to apply it verctorised?
```python
df_weights_wide * method( df_adj_zns_wide) * param
```
instead of the for-loop.